src/mlpack/methods/boostmetric/boostmetric_impl.hpp

```diff
@@ -0,0 +1,179 @@
+#define MLPACK_METHODS_BOOSTMETRIC_BOOSTMETRIC_IMPL_HPP
+
+// In case it was not already included.
+#include "boostmetric.hpp"
+
+namespace mlpack {
+namespace boostmetric {
+
+/**
+ * Takes in a reference to the dataset. Copies the data, initializes
+ * all of the member variables and constraint object.
+ */
+template<typename MetricType>
+BOOSTMETRIC<MetricType>::BOOSTMETRIC(const arma::mat& dataset,
+                                     const arma::Row<size_t>& labels,
+                                     const size_t k) :
+    dataset(dataset),
+    labels(labels),
+    k(k),
+    maxIter(500),
+    tolerance(1e-7),
+    wTolerance(1e-5),
+    constraint(dataset, labels, k)
+{ /* nothing to do */ }
+
+// Calculate inner product of two matrices.
+inline double innerProduct(arma::mat& Ar, arma::mat& Z)
+{
+  double sum = 0.0;
+
+  for (size_t i = 0; i < Z.n_elem; i++)
+    sum += Ar(i) * Z(i);
+
+  return sum;
+}
+
+template<typename MetricType>
+void BOOSTMETRIC<MetricType>::LearnDistance(arma::mat& outputMatrix)
+{
+  Timer::Start("boostmetric");
+
+  // See if we were passed an initialized matrix.
+  if ((outputMatrix.n_rows != dataset.n_rows) ||
+      (outputMatrix.n_cols != dataset.n_rows))
+  {
+    outputMatrix.eye(dataset.n_rows, dataset.n_rows);
+  }
+
+  // Compute triplets.
+  arma::Mat<size_t> triplets;
+  constraint.Triplets(triplets, dataset, labels);
+
+  size_t N = triplets.n_cols;
+  size_t dim = dataset.n_rows;
+
+  // Initialize u.
+  arma::vec u;
+  u.ones(N);
+
+  // Normalize u such that sum(u_i) = 1
+  u = u / N;
+
+  arma::cube A(dim, dim, N);
+
+  // Initialize A (coefficient matrix of difference between distance
+  // of impostors & target neighbors).
+  for (size_t i = 0; i < N; i++)
+  {
+    A.slice(i) = (dataset.col(triplets(0, i)) - dataset.col(triplets(2, i))) *
+        arma::trans(dataset.col(triplets(0, i)) - dataset.col(triplets(2, i)))
+        -
+        (dataset.col(triplets(0, i)) - dataset.col(triplets(1, i))) *
+        arma::trans(dataset.col(triplets(0, i)) - dataset.col(triplets(1, i)));
+  }
+
+  // Declare weighted sum of Ar matrices.
+  arma::mat Acap;
+
+  // Now iterate!
+  for (size_t i = 0; i < maxIter; i++)
+  {
+    // Initialize Acap.
+    Acap.zeros(dim, dim);
+    for (size_t j = 0; j < N; j++)
+      Acap += u(j) * A.slice(j);
+
+    arma::vec eigval;
+    arma::mat eigvec;
+    arma::eig_sym(eigval, eigvec, (Acap + trans(Acap) / 2));
+
+    // Get maximum eigvalue.
+    double maxEig = arma::max(eigval);
+
+    if (maxEig < tolerance)
+    {
+      Log::Info << "BoostMetric: minimized within tolerance " << tolerance
+            << "; " << "terminating optimization." << std::endl;
+      break;
+    }
+
+    // Output current maximum eigen value.
+    Log::Info << "BoostMetric: iteration " << i << ", maximum eigen value "
+        <<  maxEig << "." << std::endl;
+
+    // Get index of maximum element.
+    arma::uvec maxIndex = arma::find(eigval == maxEig);
```

Comment:
I think we can at least avoid this calculation if we have something like index_max() but armadillo supports it starting from v 7.2000